symfony/serializer version 7.0.7 now throws NotNormalizableValueException

### Symfony version(s) affected

7.0.7 

### Description

i inherited the following code, and now i am unsure wether this is a bug in the last patch-release or a bug in my code. up until version 7.0.6 it worked just fine so far.


### How to reproduce

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Trias>
    <ServiceDelivery>
        <DeliveryPayload>
            <TripResponse>
                <foo />
                <foo />
            </TripResponse>
        </DeliveryPayload>
    </ServiceDelivery>
</Trias>
```

with the following DTOs

```php
final class Trias
{
    private ?ServiceDeliveryStructure $serviceDelivery = null;

    /**
     * @return ServiceDeliveryStructure
     */
    public function getServiceDelivery()
    {
        return $this->serviceDelivery;
    }

    public function setServiceDelivery(ServiceDeliveryStructure $serviceDelivery): void
    {
        $this->serviceDelivery = $serviceDelivery;
    }
}
```

```php
final class ServiceDeliveryStructure
{
    private ?DeliveryPayloadStructure $deliveryPayload = null;

    /**
     * @return DeliveryPayloadStructure
     */
    public function getDeliveryPayload()
    {
        return $this->deliveryPayload;
    }

    public function setDeliveryPayload(DeliveryPayloadStructure $deliveryPayload): void
    {
        $this->deliveryPayload = $deliveryPayload;
    }
}
```


```php
final class DeliveryPayloadStructure
{
    private ?TripResponse $tripResponse = null;

    /**
     * @return TripResponse
     */
    public function getTripResponse()
    {
        return $this->tripResponse;
    }

    public function setTripResponse(TripResponse $tripResponse): void
    {
        $this->tripResponse = $tripResponse;
    }
}
```

starting with version 7.0.7 i'll get the following error:

```
Symfony\Component\Serializer\Exception\NotNormalizableValueException: The type of the "tripResponse" attribute for class "EosUptrade\FIS\TRIAS\Dto\DeliveryPayloadStructure" must be one of "EosUptrade\FIS\TRIAS\Dto\TripResponse" ("array" given).
```

when calling

```php
$serializer->deserialize($content, Trias::class, XmlEncoder::FORMAT);
```

with the following service configuration:

```php
    $services->set('trias.name_converter', UppercaseFirstLetterNameConverter::class);
    $services->set('trias.property_type_extractor', PhpDocExtractor::class);

    $services
        ->set('trias.get_set_method_normalizer', GetSetMethodNormalizer::class)
        ->args([
            null,
            service('trias.name_converter'),
            service('trias.property_type_extractor'),
        ]);

    $services->set('trias.array_denormalizer', ArrayDenormalizer::class);
    $services->set('trias.xml_encoder', XmlEncoder::class);

    $services
        ->set('trias.serializer', Serializer::class)
        ->public()
        ->args([
            [
                service('trias.get_set_method_normalizer'),
                service('trias.array_denormalizer'),
            ],
            [service('trias.xml_encoder')],
        ]);
```

### Possible Solution

_No response_

### Additional Context

_No response_

i created a little poc here to illustrate the issue:

https://github.com/verfriemelt-dot-org/serializer-poc

```shell
$ composer install 2>/dev/null
$ vendor/bin/phpunit
PHPUnit 11.1.3 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.3.6
Configuration: /home/easteregg/src/serialize-poc/phpunit.xml

E                                                                   1 / 1 (100%)

Time: 00:00.114, Memory: 6.00 MB

There was 1 error:

1) Test::test
Symfony\Component\Serializer\Exception\NotNormalizableValueException: The type of the "tripResponse" attribute for class "Dto\DeliveryPayloadStructure" must be one of "Dto\TripResponse" ("array" given).

/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Exception/NotNormalizableValueException.php:32
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:597
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:366
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Serializer.php:238
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:537
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:366
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Serializer.php:238
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:537
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:366
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Serializer.php:238
/home/easteregg/src/serialize-poc/vendor/symfony/serializer/Serializer.php:143
/home/easteregg/src/serialize-poc/poc/Test.php:51

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

where as if you require 7.0.6 it works:

```shell
$ composer require symfony/serializer:7.0.6 2>/dev/null
$ vendor/bin/phpunit
PHPUnit 11.1.3 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.3.6
Configuration: /home/easteregg/src/serialize-poc/phpunit.xml

.                                                                   1 / 1 (100%)

Time: 00:00.103, Memory: 6.00 MB

OK (1 test, 0 assertions)
```

this example is as small as i could create it, hope that helps :thinking:

--- a/src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php
+++ b/src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php
@@ -107,7 +107,7 @@ class GetSetMethodNormalizer extends AbstractObjectNormalizer
     {
         return !$method->isStatic()
             && (\PHP_VERSION_ID < 80000 || !$method->getAttributes(Ignore::class))
-            && 1 === $method->getNumberOfRequiredParameters()
+            && 0 < $method->getNumberOfParameters()
             && str_starts_with($method->name, 'set');
     }
 

--- a/src/Symfony/Component/Serializer/Tests/Normalizer/GetSetMethodNormalizerTest.php
+++ b/src/Symfony/Component/Serializer/Tests/Normalizer/GetSetMethodNormalizerTest.php
@@ -538,6 +538,18 @@ class GetSetMethodNormalizerTest extends TestCase
         $obj = $this->normalizer->denormalize(['foo' => 'foo'], GetSetDummyChild::class);
         $this->assertSame('foo', $obj->getFoo());
     }
+
+    /**
+     * @testWith [{"foo":"foo"}, "getFoo", "foo"]
+     *           [{"bar":"bar"}, "getBar", "bar"]
+     */
+    public function testSupportsAndDenormalizeWithOptionalSetterArgument(array $data, string $method, string $expected)
+    {
+        $this->assertTrue($this->normalizer->supportsDenormalization($data, GetSetDummyWithOptionalAndMultipleSetterArgs::class));
+
+        $obj = $this->normalizer->denormalize($data, GetSetDummyWithOptionalAndMultipleSetterArgs::class);
+        $this->assertSame($expected, $obj->$method());
+    }
 }
 
 class GetSetDummy
@@ -859,5 +871,31 @@ class GetSetDummyParent
     public function setFoo($foo)
     {
         $this->foo = $foo;
+    }
+}
+
+class GetSetDummyWithOptionalAndMultipleSetterArgs
+{
+    private $foo;
+    private $bar;
+
+    public function getFoo()
+    {
+        return $this->foo;
+    }
+
+    public function setFoo($foo = null)
+    {
+        $this->foo = $foo;
+    }
+
+    public function getBar()
+    {
+        return $this->bar;
+    }
+
+    public function setBar($bar = null, $other = true)
+    {
+        $this->bar = $bar;
     }
 }

